### PR TITLE
Extracting metadata using a file name parser

### PR DESCRIPTION
**Pull request by SlimaneAmar**
Signed-off-by: Slimane AMAR <slimane.amar@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
This PR introduces the possibility of extracting metadata from an entsoe case (uct and cgmes) by parsing the name of the file according to the entsoe codification and sending them via a notification

**What is the current behavior?** *(You can also link to an open issue here)*
No metadata extraction

**What is the new behavior (if this is a feature change)?**
Metadata extraction

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: case-server/pom.xml [hunk 1]
```diff
@@ -125,6 +130,25 @@
             <artifactId>log4j-over-slf4j</artifactId>
             <scope>runtime</scope>
         </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-triple-store-impl-rdf4j</artifactId>
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
```
Comment (geofjamg) [inline case-server/pom.xml:138]: You don't have to specify rdf4j dependencies, it is transitively pulled by powsybl-triple-store-impl-rdf4j.

File: case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java [hunk 2]
```diff
@@ -18,23 +20,40 @@
 /**
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
  */
-@AllArgsConstructor
-@NoArgsConstructor
-@Getter
+@SuperBuilder
+@Getter()
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java:24]: is () needed?

File: case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java [hunk 3]
```diff
@@ -18,23 +20,40 @@
...
-    private UUID uuid;
+    @NonNull protected String   name;
+    @NonNull protected UUID     uuid;
+    @NonNull protected String   format;
 
-    public static Message<String> getMessage(CaseInfos caseInfos) {
+    public static CaseInfos create(String fileBaseName, UUID caseUuid, String format, FileNameInfos fileNameInfos) {
+        switch (fileNameInfos.getType()) {
+            case ENTSOE:
+                EntsoeFileName entsoeFileName = (EntsoeFileName) fileNameInfos;
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java:39]: I am not sure this is the right place for entsoe related code. What about if we would like to move FileName impl in another Maven module to really leverage plugins archi?

File: case-server/src/main/java/com/powsybl/caseserver/EntsoeCaseInfos.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/EntsoeCaseInfos.java:2]: Year is incorrect

File: case-server/src/main/java/com/powsybl/caseserver/EntsoeCaseInfos.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import com.powsybl.iidm.network.Country;
+import io.swagger.annotations.ApiModel;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.experimental.SuperBuilder;
+import org.joda.time.DateTime;
+import org.springframework.messaging.Message;
+
+/**
+ * A utility class to work with Entsoe file name parsers.
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/EntsoeCaseInfos.java:19]: It is more a data class than a utility class?

File: case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileName.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileName.java:2]: Year is incorrect

File: case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileName.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+package com.powsybl.caseserver.parsers;
+
+import com.powsybl.entsoe.util.EntsoeGeographicalCode;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.experimental.SuperBuilder;
+import org.joda.time.DateTime;
+
+/**
+ * A utility class to work with Entsoe file name parsers.
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileName.java:16]: utility?

File: case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileName.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * @author Slimane Amar <slimane.amar at rte-france.com>
+ */
+@SuperBuilder
+@Getter()
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileName.java:21]: () ?

File: case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileNameParser.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileNameParser.java:2]: Year is incorrect

File: case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileNameParser.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        return df.withZone(DateTimeZone.forID("Europe/Brussels")).parseDateTime(dateStr);
+    }
+
+    @Override
+    public boolean exists(String fileBaseName) {
+        return FILE_NAME_REGEX.matcher(fileBaseName).matches();
+    }
+
+    @Override
+    public EntsoeFileName parse(String fileBaseName) {
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/EntsoeFileNameParser.java:41]: Why you did not rely on powsybl-core EntsoeFileName class? Or improve it if necessary?

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameInfos.java [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package com.powsybl.caseserver.parsers;
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameInfos.java:1]: License is missing

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameInfos.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+
+public interface FileNameInfos {
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameInfos.java:3]: author is missing

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParser.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+package com.powsybl.caseserver.parsers;
+
+/**
+ * A utility class to work with case name parsers.
+ *
+ * @author Slimane Amar <slimane.amar at rte-france.com>
+ */
+public interface FileNameParser {
+
+    boolean exists(String fileBaseName);
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParser.java:16]: Coud we find a more accurate name for this? isParseable?

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParser.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+    default FileNameInfos parse(String fileBaseName) {
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParser.java:18]: Why is it a default method? Default methods are only recommended for binary compatibility.

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsers.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
+package com.powsybl.caseserver.parsers;
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsers.java:3]: Year is incorrect

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsers.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
+package com.powsybl.caseserver.parsers;
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsers.java:1]: Licence should be before package.

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsersLoader.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (geofjamg) [inline case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsersLoader.java:2]: Year is incorrect

File: case-server/src/test/java/com/powsybl/caseserver/CaseFileNameParserTests.java [hunk 18]
```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (geofjamg) [inline case-server/src/test/java/com/powsybl/caseserver/CaseFileNameParserTests.java:2]: Year is incorrect

File: pom.xml [hunk 19]
```diff
@@ -311,6 +324,23 @@
                 <version>${slf4j.version}</version>
                 <scope>runtime</scope>
             </dependency>
+            <dependency>
+                <groupId>org.eclipse.rdf4j</groupId>
```
Comment (geofjamg) [inline pom.xml:328]: rdf4j dependencies have to be removed (and are not sync anymore with core since a recent PR)

Comment (SlimaneAmar) [reply pom.xml:328]: It is to fix the error when we get a case with xiidm format :
curl -X GET "http://localhost:5000/v1/cases/c23b830e-92aa-4804-b911-614dc541349b?xiidm=true" -H "accept: */*"
com.powsybl.commons.PowsyblException: No implementation available for triple store rdf4j

File: case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java (hunk 3 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java:39]: It is to fix the error when we get a case with xiidm format :
curl -X GET "http://localhost:5000/v1/cases/c23b830e-92aa-4804-b911-614dc541349b?xiidm=true" -H "accept: */*"
com.powsybl.commons.PowsyblException: No implementation available for triple store rdf4j

File: case-server/pom.xml (hunk 1 above)
Comment (SlimaneAmar) [reply case-server/pom.xml:138]: It is to fix the error when we get a case with xiidm format :
curl -X GET "http://localhost:5000/v1/cases/c23b830e-92aa-4804-b911-614dc541349b?xiidm=true" -H "accept: */*"
com.powsybl.commons.PowsyblException: No implementation available for triple store rdf4j

File: case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java (hunk 2 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java:24]: done

File: case-server/src/main/java/com/powsybl/caseserver/EntsoeCaseInfos.java (hunk 5 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/EntsoeCaseInfos.java:19]: Changed by this :
A class to store metatada for Entsoe file name parser

File: case-server/src/test/java/com/powsybl/caseserver/CaseFileNameParserTests.java (hunk 18 above)
Comment (SlimaneAmar) [reply case-server/src/test/java/com/powsybl/caseserver/CaseFileNameParserTests.java:2]: Done

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsersLoader.java (hunk 17 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsersLoader.java:2]: Done

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsers.java (hunk 16 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsers.java:1]: Done

File: case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsers.java (hunk 15 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/parsers/FileNameParsers.java:3]: Done

File: case-server/src/main/java/com/powsybl/caseserver/EntsoeCaseInfos.java (hunk 4 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/EntsoeCaseInfos.java:2]: Done

File: case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java (hunk 2 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java:24]: Done

**Comment by SlimaneAmar** [comment]
> I think we should also move entsoe related implementation classes in a entsoe sub package.
A  'entsoe' subpackage has been created int the package 'parser' and the package 'caseserver'

File: case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java (hunk 3 above)
Comment (SlimaneAmar) [reply case-server/src/main/java/com/powsybl/caseserver/CaseInfos.java:39]: A 'entsoe' subpackage has been created int the package 'parser' and the package 'caseserver'

**Pull request merged**